core/src/main/java/com/oath/oak/EntrySet.java

```diff
@@ -473,274 +387,137 @@ boolean casNextEntryIndex(int ei, int nextOld, int nextNew) {
+     * Returns false if:
+     *   (1) there is no such entry or
+     *   (2) entry has no key set
      * The thread-local ByteBuffer can be reused by different threads, however as long as
      * a thread is invoked the ByteBuffer is related solely to this thread.
      */
-    ByteBuffer readKey(int ei) {
+    boolean readKey(KeyBuffer key, int ei) {
         if (ei == INVALID_ENTRY_INDEX) {
-            return null;
+            key.invalidate();
+            return false;
         }
 
-        long keyReference = getKeyReference(ei);
-        return keyRefToByteBuffer(keyReference, memoryManager);
-    }
-
-    static ByteBuffer keyRefToByteBuffer(long keyRef, MemoryManager mm) {
-        int[] keyArray = UnsafeUtils.longToInts(keyRef);
-        int blockID = getKeyBlockIDFromIntArray(keyArray);
-        int keyPosition = getPositionFromIntArray(keyArray);
-        int length = getKeyLengthFromIntArray(keyArray);
-
-        return mm.getByteBufferFromBlockID(blockID, keyPosition, length);
-    }
-
-    void setKeyOutputRBuff(int ei, OakDetachedReadKeyBuffer keyRef) {
-        if (ei == INVALID_ENTRY_INDEX) {
-            return;
+        long reference = getKeyReference(ei);
+        key.setReference(reference);
+        boolean isValid = KEY.decode(key, reference);
+        if (isValid) {
+            memoryManager.readByteBuffer(key);
         }
-        long keyReference = getKeyReference(ei);
-        keyRefToOakRRef(keyReference, keyRef);
+        return isValid;
     }
 
     /**
-     * Sets the given external key reference (OakRReference) given the long key reference.
-     * Used while iterating, when lookUp context is not used.
-     * There is no copy just a special ByteBuffer for a single key.
+     * Reads a value from entry at the given entry index (from off-heap).
+     * Returns false if:
+     *   (1) there is no such entry or
+     *   (2) entry has no value set
      * The thread-local ByteBuffer can be reused by different threads, however as long as
      * a thread is invoked the ByteBuffer is related solely to this thread.
      */
-    static void keyRefToOakRRef(long keyReference, OakDetachedReadKeyBuffer oakKeyRef) {
-        int[] keyArray = UnsafeUtils.longToInts(keyReference);
-        int blockID = getKeyBlockIDFromIntArray(keyArray);
-        int keyPosition = getPositionFromIntArray(keyArray);
-        int length = getKeyLengthFromIntArray(keyArray);
-
-        oakKeyRef.setReference(blockID, keyPosition, length);
-    }
-
-    /**
-     * Sets the given value reference (OakRReference) given the entry index.
-     * Returns false if: (1)there is no such entry or (2)entry has no value set
-     * Returns true otherwise
-     * a thread is invoked on the ByteBuffer is related solely to this thread.
-     */
-    boolean setValueOutputRBuff(int ei, OakDetachedReadKeyBuffer valueRef) {
+    boolean readValue(ValueBuffer value, int ei) {
         if (ei == INVALID_ENTRY_INDEX) {
             return false;
```

Comment:
It is needed for the case that we check if the buffer is valid but we cannot use the method's return value.
We only ever validate if the key is valid since for the value we have the `ValueState`.
For correctness, however, `readValue()` should behave in the same manner.
Fixed.